Update Write Your First Application Tutorial (#1676)
hyperledger/fabric: docs/source/write_first_app.rst

  .. image:: images/AppConceptsOverview.png

  **2. Explore a sample smart contract.**

>   We’ll inspect the sample assetTransfer (javascript) smart contract to learn about the transactions within them,

Reviewer: within them -> within it (this refers to the smart contract, right?)

Author: Done, thanks for the review and comments.

>   and how they are used by applications to query and update the ledger.

Reviewer: applications -> an application
I think it's better to stick to singular because it matches the title of the tutorial.

Author: done.

  **3. Interact with the smart contract with a sample application.** Our application will
  use the assetTransfer smart contract to create, query, and update assets on the ledger.

>   We'll get into the code of the apps and the transactions they create, including initializing

Reviewer: apps -> app

Author: done.

          smart contract and application, but the ``fabric-samples`` repo also
          contains Go, Java and TypeScript versions of this sample smart contract. To try the
          Go, Java or TypeScript versions, change the ``javascript`` argument
          for ``./network.sh deployCC -ccl javascript`` below to either ``go``, ``java`` or ``typescript``
          and follow the instructions written to the terminal. You may use any chaincode language sample with
          the javascript application sample (e.g javascript application calling go chaincode functions or
          javascript application calling typescript chaincode functions, etc.)

Navigate to the ``test-network`` subdirectory within your local clone of the
``fabric-samples`` repo.

.. code:: bash

  cd fabric-samples/test-network

Launch the Fabric test network using the ``network.sh`` shell script.

.. code:: bash

  ./network.sh up createChannel -c mychannel -ca

This command will deploy the Fabric test network with two peers, three certificate authorities (Orderer, Org1, Org2),
and an ordering service. Instead of using the cryptogen tool, we bring up the test network using Certificate Authorities, 

> hence the ``-ca`` flag. Additionally, the org admin user registration was bootstrapped when the Certificate Authority was started.  

Reviewer: was -> is  twice

Author: done.

In a later step, we will show how the sample application completes the admin enrollment.

Next, let's deploy the chaincode by calling the ``./network.sh`` script with the chaincode name and language options.

.. code:: bash

  ./network.sh deployCC -ccn basic -ccl javascript

.. note:: Behind the scenes, this script uses the chaincode lifecycle to package, install,

>           query installed, approve for my org, and commit chaincode.

Reviewer 2: ```suggestion
          query installed chaincode, approve chaincode for both Org1 & Org2, and finally commit the chaincode.
```

Author: Thank you for your review, done.

        try {
            // load the network configuration
            const ccpPath = path.resolve(__dirname, '..', '..', 'test-network', 'organizations', 'peerOrganizations', 'org1.example.com', 'connection-org1.json');
            const fileExists = fs.existsSync(ccpPath);
            if (!fileExists) {
                throw new Error(`no such file or directory: ${ccpPath}`);
            }
            const ccp = JSON.parse(fs.readFileSync(ccpPath, 'utf8'));

            // Create a new file system based wallet for managing identities.
            const walletPath = path.join(__dirname, 'wallet');
            const wallet = await Wallets.newFileSystemWallet(walletPath);
            console.log(`Wallet path: ${walletPath}`);

            // Steps:
            // Note: Steps 1 & 2 need to done only once in an app-server per blockchain network
            // 1. register & enroll admin user with CA, stores admin identity in local wallet
            await enrollAdmin.EnrollAdminUser();

This command stores the CA administrator's credentials in the ``wallet`` directory.
You can find administrator's certificate and private key in the ``wallet/admin.id``
file.

> .. note:: If you decide to start over by taking down the network and bringing it back up again, you will

Reviewer 2: Great ! thanks for adding this note

        await registerUser.RegisterAppUser();

Similar to the admin enrollment, this program uses a CSR to enroll ``appUser`` and
store its credentials alongside those of ``admin`` in the wallet. We now have
identities for two separate users --- ``admin`` and ``appUser`` --- that can be
used by our application.

Scrolling further down in your terminal output, you should see confirmation of the app user registration
similar to this:

.. code:: bash

  Successfully registered and enrolled user appUser and imported it into the wallet

Third, how the sample application interacts with the chaincode
-----------------------------------------------------------

In the prior steps, the application generated the admin and app user credentials and placed them in the wallet.
If the credentials exist and have the correct permissions attributes associated with them, the sample application user
will be able to call chaincode functionality from the sample application after getting reference to the channel name and contract name.

You will notice that in the following lines of application code, the application is getting reference

> to the channel on the network and the contract name:

Reviewer 2: ```suggestion
to the Contract using the  contract name and channel name via Gateway:
```

Author: Done.

> 

Reviewer 2: add a blurb about discovery & asLocalhost

"Our connection configuration specifies only the peer from your own Org. We tell node client sdk to use the service discovery (running on the peer), which fetches other peers that are currently online, metadata like relevant endorsement policies and any static information it would have otherwise needed to communicate with the rest of the node.
 
The asLocalhost set to true, tells it to connect as localhost, since our client is running on same network as the other fabric nodes."

Author: Thanks for the reminder and the suggested text!  Done.

.. code:: bash

    //3. Prepare to call chaincode using fabric javascript node sdk
    // Create a new gateway for connecting to our peer node.
        const gateway = new Gateway();
        await gateway.connect(ccp, {
            wallet,
            identity: registerUser.ApplicationUserId,
            discovery: {enabled: true, asLocalhost: true}
        });
        try {
            // Get the network (channel) our contract is deployed to.
            const network = await gateway.getNetwork(myChannel);

            // Get the contract from the network.
            const contract = network.getContract(myChaincodeName);

Fourth, initializing the ledger with some data we can work with
---------------------------------------------------------------

> Now that we are to the point where we are actually having the sample application submit transactions, let’s

Reviewer: to the point  -> at the point

Author: Done.

> go through them in sequence where we'll show the snippet of the application code, the snippet of the corresponding

Reviewer: where we'll show the snippet of the application code, the snippet of the corresponding -> , look at the snippets of the application code and the corresponding

Author: Done.

                Size: 10,
                Owner: 'Max',
                AppraisedValue: 600,
            },
            {
                ID: 'asset5',
                Color: 'black',
                Size: 15,
                Owner: 'Adriana',
                AppraisedValue: 700,
            },
            {
                ID: 'asset6',
                Color: 'white',
                Size: 15,
                Owner: 'Michel',
                AppraisedValue: 800,
            },
        ];

        for (const asset of assets) {
            asset.docType = 'asset';
            await ctx.stub.putState(asset.ID, Buffer.from(JSON.stringify(asset)));
            console.info(`Asset ${asset.ID} initialized`);
        }
    }

The terminal output entry should look similar to below:

.. code:: bash

  Submit Transaction: InitLedger creates the initial set of assets on the ledger.

Fifth, the rest of the transaction submission sequence from the sample application
----------------------------------------------------------------------------------

First, a word about querying the ledger.

Each peer in a blockchain network hosts a copy of the `ledger <./ledger/ledger.html>`_. An application

> program can view the most recent data from the ledger using read only invocations of

Reviewer: read only -> read-only

Author: Done.

      }
      return assetJSON.toString();
  }

Terminal Output:

.. code:: bash

  ***********************
  Evaluate Transaction: AssetExists asset1
    result: true
  Submit Transaction: UpdateAsset asset1, new AppraisedValue : 350
  Evaluate Transaction: ReadAsset asset1
    result: {
    "ID": "asset1",
    "Color": "blue",
    "Size": "5",
    "Owner": "Tomoko",
    "AppraisedValue": "350"
  }

In this part of the sequence, the sample application attempts to submit
a ``'UdpateAsset'`` transaction for an asset that we know does not exist (``asset70``).
We expect that we will get an error because you cannot update an asset that does not exist,
which is why it is a good idea to check if an asset exists prior to attempting an
asset update or deletion.

Sample application -- ``'UpdateAsset'``

.. code:: bash

   try {
        console.log('\nSubmit Transaction: UpdateAsset asset70');
        //Non existing asset asset70 should throw Error
        await contract.submitTransaction('UpdateAsset', 'asset70', 'blue', '5', 'Tomoko', '300');
    } catch (error) {
        console.log(`Expected an error on UpdateAsset of non-existing Asset: ${error}`);
    }

> Chaincode -- ``'UdpateAsset'``

Reviewer: UpdateAsset

Author: Done.

which is the result that was returned here.
The application starts by bringing in scope two key classes from the
``fabric-network`` module; ``Wallets`` and ``Gateway``. These classes
will be used to locate the ``appUser`` identity in the wallet, and use it to
connect to the network:

We can use the ``queryCar`` transaction to query against any car, using its
key (e.g. ``CAR0``) and get whatever make, model, color, and owner correspond to
that car.
.. code:: bash

Great. At this point you should be comfortable with the basic query transactions
in the smart contract and the handful of parameters in the query program.
  const { Gateway, Wallets } = require('fabric-network');

Time to update the ledger...
First, the program uses the Wallet class to get our application user from our file system.

Updating the ledger
-------------------
.. code:: bash

Now that we’ve done a few ledger queries and added a bit of code, we’re ready to
update the ledger. There are a lot of potential updates we could make, but
let's start by creating a **new** car.
  const identity = await wallet.get(registerUser.ApplicationUserId);

> From an application perspective, updating the ledger is simple. An application

Reviewer 2: I think this para about Update to the ledger got lost in the new description. Seem like a good detail to keep along with the tutorial/write_first_app.diagram.2.png diagram

Author: Agreed, re-added paragraph and image.

keep the ledger consistent across the network is called consensus, and it is
explained in detail in this `section <./peers/peers.html>`_.

> The ``evaluateTransaction`` method represents one of the simplest interactions

Reviewer 2: if possible, move this para about evaluateTransaction before the description about `submitTransaction` in line 859, to describe the simpler evaluateTransaction first as a natural progression

Author: Agree, it flows better this way.  Done.

with a smart contract in blockchain network. It simply picks a peer defined in
the connection profile and sends the request to it, where it is evaluated. The
smart contract queries all the cars on the peer's copy of the ledger and returns

> the result to the application. This interaction does not result in an update the

Reviewer: and update the -> an update of the

Author: Done.

ledger.

The asset-transfer ('basic') smart contract
-------------------------

> The smart contract sample is available in following languages:

Reviewer: in following languages -> in the following languages

Author: Done.
